### ISSUE TITLE
NCBI XLS BioSample consistency with tsv file

### ISSUE THREAD

**Issue by laserson** [airr-community/airr-standards#20]
The Travis test does not check for consistency with the NCBI files.  It seems there is quite a lot missing in those files compared with the tsv file, which I am guessing is by design?  It would still be worthwhile to put in an automated test to make sure that if something is added, it doesn't also need to be reflected in the NCBI templates.  This might mean annotating each field with whether or not it should be in the NCBI template and checking for it.

**Comment by BusseChristian** [comment]
Yes, the NCBI template only represent certain sections of the tsv. We are still in process of fixing the xls versions and therefore have not yet gotten to the XML versions. In addition, the are some keyword that are
required by NCBI so they have to be harmonized or mapped. Is there a possibility to use the Python script also to check the xls files?

**Comment by laserson** [comment]
No problem to read the xls files with Python.  I just don't know what it is we should be checking.

**Comment by BusseChristian** [comment]
For the BioSample XLS Template the first non-comment line would have to be compared with the keywords of set 1 (subject, diag. & intervent.), 2 and 3 (cell. processing). Here is some pseudo-code:

```
open "AIRR_Minimal_Standard_Data_Elements.tsv"
while read_line do
  next if split(read_line, "\t")[1] is not in
    ("1 / subject", "1 / diag. & intervent.", "2 / sample", "3 / process (cell)")
  push split(read_line, "\t")[5] to table_selected
done

open "NCBI_implementation/NCBI Templates/AIRR_BioSample_v1.0.xls"
if get(number of sheets) > 1
  exit with error "too many sheets"
else
  open sheet_1
  while read_line do
    next if match_RE_extended(read_line, /^\s*#.*/)   # Ignore lines starting with a hash/pound
    for each current_cell in split_line2cells(read_line) do
      current_cell := replace(current_cell, "^\*", "")   # Remove leading asterisk in mandatory elements
      push current_cell to array_keywords
    done
    break   # Read only the first non-comment line
  done
endif

compare set(array_keywords), set(table_selected)
```

**Comment by laserson** [comment]
@BusseChristian I implemented your logic in the testing script (which is merged into development).  Here are the results:

```
link_type                      is found in MiAIRR table tsv but not in NCBI Biosample template xls
linked_subjects                is found in MiAIRR table tsv but not in NCBI Biosample template xls
disease_stage                  is found in MiAIRR table tsv but not in NCBI Biosample template xls
prior_therapies                is found in MiAIRR table tsv but not in NCBI Biosample template xls
synthetic                      is found in MiAIRR table tsv but not in NCBI Biosample template xls
age_event                      is found in MiAIRR table tsv but not in NCBI Biosample template xls
study_group_description        is found in MiAIRR table tsv but not in NCBI Biosample template xls
cells_per_reaction             is found in MiAIRR table tsv but not in NCBI Biosample template xls
subject_id                     is found in MiAIRR table tsv but not in NCBI Biosample template xls
immunogen                      is found in MiAIRR table tsv but not in NCBI Biosample template xls
disease_diagnosis              is found in MiAIRR table tsv but not in NCBI Biosample template xls
ancestry_population            is found in MiAIRR table tsv but not in NCBI Biosample template xls
cell_phenotype                 is found in MiAIRR table tsv but not in NCBI Biosample template xls
sample_id                      is found in MiAIRR table tsv but not in NCBI Biosample template xls
collection_date                is found in MiAIRR table tsv but not in NCBI Biosample template xls
intervention                   is found in MiAIRR table tsv but not in NCBI Biosample template xls
race                           is found in MiAIRR table tsv but not in NCBI Biosample template xls
disease_length                 is found in MiAIRR table tsv but not in NCBI Biosample template xls
strain_name                    is found in MiAIRR table tsv but not in NCBI Biosample template xls
cell_processing_protocol       is found in MiAIRR table tsv but not in NCBI Biosample template xls
medical_history                is found in MiAIRR table tsv but not in NCBI Biosample template xls
ethnicity                      is found in MiAIRR table tsv but not in NCBI Biosample template xls
single_cell                    is found in MiAIRR table tsv but not in NCBI Biosample template xls
cell_involved                  is found in NCBI Biosample template xls but not in MiAIRR table tsv
cell_subset_phenotype          is found in NCBI Biosample template xls but not in MiAIRR table tsv
processing_protocol            is found in NCBI Biosample template xls but not in MiAIRR table tsv
sample_name                    is found in NCBI Biosample template xls but not in MiAIRR table tsv
isolate                        is found in NCBI Biosample template xls but not in MiAIRR table tsv
sample_collection_time         is found in NCBI Biosample template xls but not in MiAIRR table tsv
```

**Comment by BusseChristian** [comment]
Most of those are resolved by commits 8f63c1e8dde61ba771bc062f1fa74f74c061ac25, 2cfaf9e85c325c34f6bba986d4daebe1cf1ee3f7, be33a6ef923b6287f63d835329c325aa3a4aa04f and 3a7c38bce47873f8bb6989e85b4e9945aa3ea8fb. The ones that remain require mapping via the new `mapping_MiAIRR_BioSample.tsv` table. Sorry for complicating things, but this way everything will be clearly documented.

**Comment by schristley** [comment]
Most of the discussion and changes where related to BioSample XLS and TSV consistency, so I've renamed the issue. Will create new issues for XML and other XLS.
